### PR TITLE
Bugfix open FileAnnotation

### PR DESCRIPTION
**Pull request by dominikl**
Fixes [Ticket 12028](https://trac.openmicroscopy.org/ome/ticket/12028) and some other issues with "Open with" functionality
- Removes the "View" button (doesn't have a function anymore)
- Fixes a bug with respect to "Open with" file attachments
- Fixes the "View" issues in Import and Script dialog on Linux 
- End the "Open with" activity if an error occurs
- Fixes a problem with mimetype detection and mimetype-application association

Test (on multiple platforms if possible, but focus on Linux)
- Attach a File to an image (dataset, whatever). Select the image and make sure that there is no "View" button anymore in the attachment popup menu (right hand side metadata panel)
- Open the Attachments view in the left hand side panel. Select an attachment and open it with "Open with...". Check that the file type <-> application association is stored, i. e. if you open the "Open with" menu again on the same file type, your previously chosen application is listed in the sub menu.
- Select an Image, click on the Publishing button in metadata panel toolbar to open the Script dialog for any script; on the popup menu at the bottom left select "View". The script should be opened in your default desktop text editor
- Import an image. In the Importer dialog, click on "View" -> "Import Log"; the import log should open in your default desktop text editor
- Select a big image (to provoke an error), select "Open with". The activity should start and immediately end with an error (before: the activity status kept on spinning).

--no-rebase

### CODE REVIEW

File: components/insight/SRC/org/openmicroscopy/shoola/env/data/model/appdata/LinuxApplicationDataExtractor.java [hunk 1]
```diff
@@ -80,26 +80,17 @@ public ApplicationData extractAppData(File file) throws Exception {
...
+    /**
+     * @param location
+     *            the location pointing to the file to be opened
+     * @returns the command string to launch the default application for the
+     *          file specified by {@code location} on Gnome based desktop will
+     *          use gnome-open and on KDE based will use kde-open
+     */
+    public String[] getDefaultOpenCommandFor(URL location) {
+        String[] defaultCommands = new String[] { "xdg-open",
+                location.toString() };
```
Comment (mtbc) [inline components/insight/SRC/org/openmicroscopy/shoola/env/data/model/appdata/LinuxApplicationDataExtractor.java]: Good idea, stands a good chance of working.

File: components/insight/SRC/org/openmicroscopy/shoola/env/data/model/appdata/LinuxApplicationDataExtractor.java [hunk 2]
```diff
@@ -80,26 +80,17 @@ public ApplicationData extractAppData(File file) throws Exception {
...
-		}
-
-		return defaultCommands;
-	}
+    /**
+     * @param location
+     *            the location pointing to the file to be opened
+     * @returns the command string to launch the default application for the
+     *          file specified by {@code location} on Gnome based desktop will
+     *          use gnome-open and on KDE based will use kde-open
```
Comment (mtbc) [inline components/insight/SRC/org/openmicroscopy/shoola/env/data/model/appdata/LinuxApplicationDataExtractor.java]: no it won't

**Comment by dominikl** [comment]
Thanks @mtbc , always forgot to check the comments...

**Comment by gusferguson** [comment]
@dominikl @pwalczysko 

Tested with OMERO.insight-5.1.0-116-9075c2b-ice34-b463-linux user-3 read-only-1.

All tests behaved as expected.
Linux is good to go when others tested.

**Comment by pwalczysko** [comment]
All behaves as expected on Mac.

**Comment by pwalczysko** [comment]
All behaves as expected on Windows 7 except the ``View`` option in the Script menu. 
There are two issues
  - the menu is badly formatted (the ``Download`` and ``View`` options are in some small extra boxes
  - the ``View`` option tries (unsuccessfully) to execute the script instead of opening it in text editor -> this option on Win7 is thus pointless

**Comment by pwalczysko** [comment]
Except; the ``View`` problem on Win7 it is all as expected here.

Nevertheless, something which is connected with this PR only indirectly, @jburel @gusferguson @dominikl : We just lost a nice peace of functionality, the View button in the right-hand pane on Attachments - expected ? Mishap ? I think this was quite useful bit, why is it gone ?

The Open with workflow on ``dv`` images ends up with 

```
java.lang.Exception: org.openmicroscopy.shoola.env.data.DSAccessException: Cannot export the image as an OME-TIFF
	at org.openmicroscopy.shoola.env.data.OMEROGateway.exportImageAsOMEObject(OMEROGateway.java:7339)
	at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.exportImageAsOMEFormat(OmeroImageServiceImpl.java:1486)
	at org.openmicroscopy.shoola.env.data.views.calls.ExportLoader$1.doCall(ExportLoader.java:87)
	at org.openmicroscopy.shoola.env.data.views.BatchCall.doStep(BatchCall.java:144)
	at org.openmicroscopy.shoola.util.concur.tasks.CompositeTask.doStep(CompositeTask.java:226)
	at org.openmicroscopy.shoola.env.data.views.CompositeBatchCall.doStep(CompositeBatchCall.java:126)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.exec(ExecCommand.java:165)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.run(ExecCommand.java:276)
	at org.openmicroscopy.shoola.util.concur.tasks.AsyncProcessor$Runner.run(AsyncProcessor.java:91)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.FileNotFoundException: /var/folders/5h/4b7kss151v74c0_jgtbt69jh0000gn/T/home/hudson/for_import_to_howe/archived-imoprt-case/user-4/read-write-1/no_projects/archivedv/CFPNEAT01_R3D.dv.log9407.ome.tiff (No such file or directory)
	at java.io.FileOutputStream.open(Native Method)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
	at org.openmicroscopy.shoola.env.data.OMEROGateway.exportImageAsOMEObject(OMEROGateway.java:7307)
	... 9 more

	at org.openmicroscopy.shoola.env.ui.UserNotifierImpl.showErrorDialog(UserNotifierImpl.java:191)
	at org.openmicroscopy.shoola.env.ui.UserNotifierImpl.notifyError(UserNotifierImpl.java:291)
	at org.openmicroscopy.shoola.env.ui.UserNotifierImpl.notifyError(UserNotifierImpl.java:259)
	at org.openmicroscopy.shoola.env.ui.ActivityComponent.showException(ActivityComponent.java:413)
	at org.openmicroscopy.shoola.env.ui.ActivityComponent.actionPerformed(ActivityComponent.java:934)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2346)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
	at java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:289)
	at java.awt.Component.processMouseEvent(Component.java:6525)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3322)
	at java.awt.Component.processEvent(Component.java:6290)
	at java.awt.Container.processEvent(Container.java:2234)
	at java.awt.Component.dispatchEventImpl(Component.java:4881)
	at java.awt.Container.dispatchEventImpl(Container.java:2292)
	at java.awt.Component.dispatchEvent(Component.java:4703)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4898)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4533)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4462)
	at java.awt.Container.dispatchEventImpl(Container.java:2278)
	at java.awt.Window.dispatchEventImpl(Window.java:2739)
	at java.awt.Component.dispatchEvent(Component.java:4703)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:751)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:702)
	at java.awt.EventQueue$3.run(EventQueue.java:696)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.awt.EventQueue$4.run(EventQueue.java:724)
	at java.awt.EventQueue$4.run(EventQueue.java:722)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:721)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```
Expected ?
Should I file a ticket ?

**Comment by dominikl** [comment]
The problem with the dv is not expected, I'll check again what's going on there.

**Comment by jburel** [comment]
Any progress @dominikl?

**Comment by dominikl** [comment]
Tested it again. Works for newly imported dv files, looks like there's something wrong with the existing dvs in archivedvs and non-archivedvs.

**Comment by pwalczysko** [comment]
@dominikl : There should be nothing wrong with the files in question. The archived and non-archived dvs are 4.4 data which have been upgraded - if this is the root of the problem, then we would have a problem with all 4.4 data.

**Comment by dominikl** [comment]
But it's not caused by this PR, same happens without the PR (with the difference that you can't see the error message because the activity is never stopped - this is fixed within this PR).

**Comment by dominikl** [comment]
Had to rebase to current develop. 
Reviewing of the last commit should be enough: Select an image, open the FigureDialog via the publishing button. Make sure the download and view entry within the scripts dropdown menu (bottom left) is displayed correctly.
@pwalczysko Not going to tackle the dv problem within this PR; not caused by this PR and just affects the existing archivedv and non-archivedv images.

**Comment by dominikl** [comment]
@pwalczysko The last commit is just a suggestion. It adds an "Open with..." menu (equivalent to the "Open with..." menu in the popup menu of the left hand side "Attachments" tab) to the metadata panel's FileAnnotation component's popup menu as replacement for the removed "View" menu item. This get's more and more complicated to explain, @pwalczysko let me know before you review this and I'll show you what I mean. - If this is not a solution we want to go for, I'll just revert the last commit.

@jburel In the last commit I make a reference to the `TreeViewerFactory` from the `DocComponent`, which is actually part of the MetadataViewerAgent package, to look up and store Application<->MimeType relations. Are these kinds of references ok? Or should this functionality moved to a neutral package, independent of any "Agents"?

**Comment by pwalczysko** [comment]
As discussed with @dominikl , the Open With menu item in the right-hand pane is going to go away now, because it does not do what the ``View`` button used to do, and this is to recognize without any dialogs or browsing the right app to open the attachment with on your local machine (Mac or Windows) and open it. This was working fine on a 5.0.8 server. This is the piece of funcitonality I think we have silently lost.
The Open With item does not deliver this functionality, offering you dialogs to browse, not working properly on Mac and is unusable on Windows, because you have to choose from ``Program Files``. So we proposed to remove the ``Open with`` from the right hand pane here and Dom will try to sort out the ``View`` problem in another PR.

The issue with the Script ``View`` button is sorted out as expected.
After removal of the ``Open With`` option from the right-hand pane attachment submenu this PR is ready to merge.

**Comment by dominikl** [comment]
Removed my last commit, so that this PR could be merged. I will bring back the View button (for file annotations in the metadata panel) in separate PR as it's not directly related to this PR.

**Pull request merged**